### PR TITLE
Remove EOL versions

### PR DESCRIPTION
**Pull request by calumy**
This PR removes the following end-of-life versions of [Python](https://devguide.python.org/versions/) and [Django](https://jefftriplett.com/django-release-cycle/): 
- Python 3.6 - EOL: 23rd December 2021
- Python 3.7 - EOL: 27th June 2023
- Django 3.2 - EOL:  April 2024 

In addition to removing the EOL versions, the [Django upgrade pre-commit hook](https://github.com/adamchainz/django-upgrade) has also been added, and fixes are included in this PR. These changes relate to:
- The use of `USE_L10N`, which was deprecated in Django 4.0: https://github.com/adamchainz/django-upgrade?tab=readme-ov-file#use_l10n
- The use of `re_path`, which has been rewritten using the `path()` syntax added in Django 2.0: https://github.com/adamchainz/django-upgrade?tab=readme-ov-file#urls

### CODE REVIEW

**Review by johnraz — CHANGES_REQUESTED** [review]
I was thinking we could merge this after the next release, like that we can issue a warning that this will be the last release to support all those older versions.

**Comment by calumy** [comment]
Drafting until #342 is merged and a new version released.

**Review by johnraz — APPROVED** [review]
LGTM

**Comment by giovannicimolin** [comment]
@calumy Thanks for your contribution! :)

I'll update the changelogs and documentation of the before the next release and be sure to add a note.

**Pull request merged**
